### ISSUE TITLE
Laguerre quadrature

### ISSUE THREAD

**Issue by Godisemo** [ajt60gaibb/FastGaussQuadrature.jl#17]
Any plans on adding code for computing Gauss–Laguerre quadrature points and weights?

**Comment by ajt60gaibb** [comment]
I do not know a reliable fast `O(n)` algorithm for Gauss--Laguerre quadrature.  Do you?

**Comment by dlfivefifty** [comment]
Would it be more accurate to say: "the algorithm could be adapted to Laguerre quadrature, but we don't know the necessary data to  implement it"?

**Comment by ajt60gaibb** [comment]
Yes, that is more accurate.  The fast Gauss quadrature code is based on two things: 

1. Relatively accuracy initial guesses, 
2. Very accurate asymptotic expansions for Laguerre polynomials,
3. A Newton iteration. 

I believe that 1) is known in the literature. But, I am not aware of good enough asymptotic expansions for Laguerre polynomial to fill the shoes for 2). Newton iteration in 3) is fine.

**Comment by Godisemo** [comment]
I don't know anything about the algorithms you use. What do you mean with asymptotic expansions in point 2?

**Comment by ajt60gaibb** [comment]
The algorithms (expect for GaussLegendre) require a fast evaluation scheme for the associated orthogonal polynomial.   For example, Gauss-Jacobi for point 2 can be http://dlmf.nist.gov/18.15#E6.  This means that a Jacobi polynomial can be evaluated much faster than using the three-term recurrence. 

For the complete algorithmic details, see http://epubs.siam.org/doi/pdf/10.1137/120889873. 

For Laguerre, see http://dlmf.nist.gov/18.15#iv.   But I am not sure if we know enough terms in the asymptotic expansions to make this a viable fast evaluation scheme.

**Comment by dholladay00** [comment]
I see in reference 5 in the readme, 2nd paragraph of the conclusions section that "Presumably, extensions of these ideas can be made to the Laguerre weights".

Does the statement "Again, we expect the asymptotic expansion to be useful in
the classical case of V (x) = −x..." essentially mean that they need more terms in the asymptotic expansion?

This is a topic I have been curious about myself since summer 2015 but do not have the mathematical knowledge to understand what is going on.

**Comment by ajt60gaibb** [comment]
Yes, I believe that asymptotic expansions for Laguerre polynomials are not sufficiently mature. 
I have a summer student (starting in June 2016) that may work on exactly this. So, I am hoping for significant progress very soon.

**Comment by dholladay00** [comment]
Just connected the dots that you are on that paper!

Literally could not have asked a better person, wow. 

I look forward to hearing of the progress.

**Comment by ajt60gaibb** [comment]
Yes, two of the authors from that paper happen to be replying to this tread. :+1:  Meanwhile, if you have any significant ideas then I would be glad to discuss. At the moment it may not even be that hard as we have not invested that much time into the Laguerre case.

I see that you are in the Department of Nuclear Engineering so I am very interested to hear what you are using Gauss quadrature for.

**Comment by dholladay00** [comment]
I stand corrected, I did not know at the time that I had come to **the** resource for quadratures, not that I'm complaining. 

There are several uses I (and the Nuclear Engineering community) have for various quadratures.

1. In my master's work, I used quadratures to calculate "semi-analytic" solutions of thermal radiative transfer (used Gauss Legendre). 
2. Quadrature sets for the solid angle discretization in the discrete ordinates formulation of the linear Boltzmann transport equation (neutron transport). 
  * A usual choice is for a Gauss-Legendre quadrature in the polar angle and a Gauss-Chebyshev quadrature in the azimuthal angle. 
  * Other quadratures are under active research such as adaptive quadrature methods, but my knowledge on that area is limited. 
3. In my current research, I need to compute photo-ionization rates in a plasma, those integrals are perfectly suited for Gauss-Leguerre quadratures. In a given simulation, there could potentially be billions of these integrals to be done. There are some adaptive methods we are looking at, but I think this could be a better solution, especially if we are no longer constrained by having to precompute and store the quadrature set(s).
